libraries/appbase/include/appbase/plugin.hpp

```diff
@@ -17,58 +17,151 @@
-         virtual void plugin_for_each_dependency(plugin_processor&& processor) = 0;
-
-         /** Abstract method to be reimplemented in final plugin implementation.
-             It is a part of initialization process triggerred by main application.
-         */
-         virtual void plugin_initialize( const variables_map& options ) = 0;
-         /** Abstract method to be reimplemented in final plugin implementation.
-             It is a part of startup process triggerred by main application.
-         */
-         virtual void plugin_startup() = 0;
-         /** Abstract method to be reimplemented in final plugin implementation.
-             It is a part of shutdown process triggerred by main application.
-         */
-         virtual void plugin_shutdown() = 0;
+using boost::program_options::options_description;
+using boost::program_options::variables_map;
+using std::string;
+using std::vector;
+using std::map;
 
+class application;
+
+class abstract_plugin : public std::enable_shared_from_this<abstract_plugin>
+{
+public:
+   enum state {
+      registered, ///< the plugin is constructed but doesn't do anything
+      initialized, ///< the plugin has initlaized any state required but is idle
+      started, ///< the plugin is actively running
+      stopped ///< the plugin is no longer running
    };
 
-   template<typename Impl>
-   class plugin;
-}
+   virtual ~abstract_plugin(){}
+
+   virtual state get_state()const = 0;
+   virtual const std::string& get_name()const  = 0;
+   static void set_program_options( options_description& cli, options_description& cfg ){};
+   virtual void register_dependencies() = 0;
+   virtual void initialize(const variables_map& options) = 0;
+   virtual void startup() = 0;
+   virtual void shutdown() = 0;
+   void set_app(application& my_app) { _app = &my_app; }
+   application* app() { return _app; }
+
+protected:
+   typedef std::function<void(abstract_plugin&)> plugin_processor;
+
+   /** Abstract method to be reimplemented in final plugin implementation.
+       It is a part of initialization/startup process triggerred by main application.
+       Allows to process all plugins, this one depends on.
+   */
+   virtual void plugin_for_each_dependency(plugin_processor&& processor) = 0;
+
+   /** Abstract method to be reimplemented in final plugin implementation.
+       It is a part of initialization process triggerred by main application.
+   */
+   virtual void plugin_initialize( const variables_map& options ) = 0;
+   /** Abstract method to be reimplemented in final plugin implementation.
+       It is a part of startup process triggerred by main application.
+   */
+   virtual void plugin_startup() = 0;
+   /** Abstract method to be reimplemented in final plugin implementation.
+       It is a part of shutdown process triggerred by main application.
+   */
+   virtual void plugin_shutdown() = 0;
+
+   void notify_app_initialize();
+   void notify_app_startup();
+private:
+   application* _app;
```

Comment:
smart pointer instead of raw ?

Reply:
ok i get what you did here. You could also have shared_ptr<application> in this internal map but it might be unnecessary overhead... Maybe some comment would be nice so everyone knows that this is not standard pointer with dynamically allocated memory...